### PR TITLE
Source Hubspot: Some incremental CRM objects and engagements

### PR DESCRIPTION
**Pull request by lgomezm**
## What
Adds a more efficient support for incremental updates on some Hubspot streams:
* Companies
* Contacts
* Deals
* Engagements

## How
For CRM objects, it uses the [CRM search endpoints](https://developers.hubspot.com/docs/api/crm/search), which allow filtering by different properties. In our case, we filter by the respective "last modified date".

For the engagements case, it uses the [Get recent engagements endpoint](https://legacydocs.hubspot.com/docs/methods/engagements/get-recent-engagements).

## Recommended reading order
1. `client.py`
2. `api.py`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>

#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

---
This change is [Reviewable](https://reviewable.io/reviews/airbytehq/airbyte/8887)

### CODE REVIEW

**Comment by lgomezm** [comment]
Related issue: #8344

**Comment by alafanechere** [comment]
Hi @lgomezm thank you very much for your contribution! Feel free to ping us when your PR is ready for review. Please also check our [doc](https://docs.airbyte.io/connector-development#publishing-a-connector) about publishing a new connector version. You basically have to bump the connector version in the `Dockerfile` + `airbyte-config/init/src/main/resources/seed/source_definitions.yaml`

**Comment by lgomezm** [comment]
Hi @alafanechere. I've just pushed some changes and bumped the version where you indicated. Also moved the PR to "Ready for review". Please let me know if there's anything I should provide on my side.

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py
Comment (alafanechere) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:231]: As you remove these `hs_time_in` fields from ignored fields in `acceptance-test-config`?

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py
Comment (alafanechere) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:553]: nit:
```suggestion
return f"/crm/v3/objects/{self.entity}/search" if self.state else f"/crm/v3/objects/{self.entity}"
```

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py
Comment (alafanechere) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:566]:
```suggestion
        self.entity = entity
        self.last_modified_field = last_modified_field
        self.associations = associations
```
I think you did not set these class variables.  `CRMObjectStream` does.

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py
Comment (alafanechere) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:569]: I know other classes are already defining a `list` method but this is a reserved python word.

Comment (alafanechere) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:569]: Moreover, this function looks like the one defined in `CRMObjectStream`. Do you think you could make them share a parent class?

**Review by alafanechere — CHANGES_REQUESTED** [review]
Hey @lgomezm, I did a first quick review. I request changes because our integration tests are not passing (a simple unused import problem for the moment): 
```
> Task :airbyte-integrations:connectors:source-hubspot:flakeCheck FAILED
[python] .venv/bin/python -m flake8 . --config /actions-runner/_work/airbyte/airbyte/tools/python/.flake8
	 ./source_hubspot/client.py:11:1: F401 'source_hubspot.api.CRMObjectStream' imported but unused
```

Could you please make sure to make the acceptance test pass? I'll also ask for another reviewer because I'm not very familiar with this connector, which has a different implementation than the classic CDK-created connectors.

@lazebnyi I allowed myself to request a review from your side because git-blame showed me you are pretty active on this connector 😄 .

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py
Comment (grubberr) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:231]: We move it back and forth
As for me we don't need to remove data `hs_*` fields from response
Some customers complained about this

https://github.com/airbytehq/airbyte/issues/8055

**Comment by lgomezm** [comment]
Hi @alafanechere. I've been updating the code as per your feedback. Here's the acceptance test results:
Screen Shot 2022-01-03 at 3 01 06 PM

Please let me know if you have any other comment.

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py
Comment (lgomezm) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:567]: The `list` function is actually being called when the connector's `read` command is invoked. I think it's called internally, so even though it does not look good because it's a reserved word, I don't know if we can rename it.

**Review by bazarnov — APPROVED** [review]

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py
Comment (grubberr) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:231]: doc string of reverted back `_filter_dynamic_fields` function points on issue #2397
which "says" that we don't need to filter-out `hs_time_*` fields now
we DO filter-out this fileds on `acceptance-test` level

**Review by grubberr — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py
Comment (lgomezm) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py]: I found where it was being invoked, so I renamed it in 42b1f42. I'm sorry I didn't know much about how these methods connected with the framework.

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py
Comment (lgomezm) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py]: I've removed this method again in 42b1f42 and can confirm it unit tests and acceptance tests still pass.

**Comment by lgomezm** [comment]
@alafanechere @grubberr I think I have addressed all your comments so far. Please take a look again and let me know if there's anything else I should update.

**Review by grubberr — APPROVED** [review]

**Comment by grubberr** [comment]
@lgomezm I have approved from my side but please get approval from  rest of team because your change pretty big
Thanks!

**Comment by lgomezm** [comment]
> @lgomezm I have approved from my side but please get approval from rest of team because your change pretty big Thanks!

@alafanechere @lazebnyi This PR is ready for a second look. Please let me know of any comment!

**Comment by lgomezm** [comment]
@alafanechere @lazebnyi Hi guys. This PR was updated a while ago. Please take a look at it when you get a chance.

**Review by lazebnyi — APPROVED** [review]

**Review by alafanechere — CHANGES_REQUESTED** [review]
Hi @lgomezm, thank for the changes you made. Could you please: 
* bump the version in this file: `airbyte-config/init/src/main/resources/config/STANDARD_SOURCE_DEFINITION/36c891d9-4bd9-43ac-bad2-10e12756272c.json` ?
* Fix the conflict with `master`
Once this is done and if acceptance tests are passing I'll take care of publishing the connector and merging the branch.
It'd be great if you could give me edit permission on this branch because I need to update the `source_specs.yaml` file after I publish the connector (by running `./gradlew airbyte-config:init:processResources`).

**Comment by lgomezm** [comment]
> Hi @lgomezm, thank for the changes you made. Could you please:
> 
> * bump the version in this file: `airbyte-config/init/src/main/resources/config/STANDARD_SOURCE_DEFINITION/36c891d9-4bd9-43ac-bad2-10e12756272c.json` ?
> * Fix the conflict with `master`
>   Once this is done and if acceptance tests are passing I'll take care of publishing the connector and merging the branch.
>   It'd be great if you could give me edit permission on this branch because I need to update the `source_specs.yaml` file after I publish the connector (by running `./gradlew airbyte-config:init:processResources`).

Hi @alafanechere I've just bumped the version as you mentioned, as well as fixed the conflicts with `master`. I'll try to figure out your branch permission request, because I am not the owner of the forked repo.

**Comment by lgomezm** [comment]
@alafanechere BTW, here are the acceptance tests still passing:
Screen Shot 2022-01-11 at 9 59 41 AM

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/client.py
Comment (alafanechere) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/client.py:16]: Duplicate import. Please run `./gradlew format`

**Review by alafanechere — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/client.py
Comment (lgomezm) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/client.py]: Done in 719b678.

**Comment by alafanechere** [comment]
>  I validated with my team and we believe our fork should already be public. Please let me know if you can't edit my branch.
The fork is indeed public but I'm not allowed to push on your fork:
```bash
git push calixa-io augustin/publish-8887:lgomez/hubspot_incremental_crm_objects
remote: Permission to calixa-io/airbyte.git denied to alafanechere.
fatal: unable to access 'https://github.com/calixa-io/airbyte.git/': The requested URL returned error: 403
```

I guess I need to be invited by your organization as a maintainer of your Airbyte repo. If it's too cumbersome please cherry-pick this commit https://github.com/airbytehq/airbyte/pull/9414/commits/485c1e7dee900638e2cae92598eb121d3f1ca7ff I made on my side branch (some other bumps were required). I'll then publish the connector and ask you to retrieve a new commit that I'll make post-publication.

**Comment by lgomezm** [comment]
Hi @alafanechere. I have just cherry-picked your commit. You can now proceed with publishing the connector.

**Comment by alafanechere** [comment]
@lgomezm thanks for your patience, I see there are some conflicts with master, could you please fix them?

**Comment by lgomezm** [comment]
Hi @alafanechere. Conflicts have been fixed and here's the acceptance test results:
Screen Shot 2022-01-13 at 2 40 52 PM

**Comment by lgomezm** [comment]
@alafanechere I've fixed conflicts again and cherry-picked your commits. Again, here's the acceptance test results:
Screen Shot 2022-01-14 at 6 23 35 AM

**Comment by lgomezm** [comment]
Hi @bazarnov. Do you know why acceptance tests are failing on this branch? The error happens when it tries to get data from the `form_submissions` stream: 
```
docker.errors.ContainerError: Command 'read --config tap_config.json --catalog catalog.json' in image '<Image: 'airbyte/source-hubspot:dev'>' returned non-zero exit status 1: ***"type": "LOG", "log": ***"level": "FATAL", "message": "404 Client Error: Not Found for url: https://api.hubapi.com/form-integrations/v1/submissions/forms?hapikey=98723888-a72e-40f8-8b71-157f0a9c1587&limit=50\nTraceback ...
```
However, it builds on master. I'm pretty sure I am not updating the `form_submissions` stream in this PR, so I am kind of lost with this one.

**Comment by marcosmarxm** [comment]
@lgomezm the `form_submission` stream was added this week but was added as an empty stream. Even after updating the acceptance test yaml file is failing? Sorry to ask again but looks there are some conflicts?

**Comment by lgomezm** [comment]
@marcosmarxm I think I've found and fixed the issue with acceptance tests in 9131635. Please give it a try again.

**Review by alafanechere — APPROVED** [review]
We finally made it @lgomezm, thank you for your contribution and patience!

**Pull request merged**
